shared/AppInsightsCommon/src/Offline.ts

```diff
@@ -0,0 +1,187 @@
+    mergeEvtNamespace
+} from "@microsoft/applicationinsights-core-js";
+
+/**
+ * this is the callback that will be called when the network status changes
+ * @param onlineState this is the current network running state
+ */
+export type OfflineCallback = (onlineState: IOfflineState) => void;
+
+/**
+ * This is the enum for the different network states current ai is experiencing
+ */
+export const enum eOfflineValue {
+    Unknown = 0,
+    Online = 1,
+    Offline = 2
+}
+/**
+ * This is the interface for the Offline state
+ * runtime state is retrieved from the browser state
+ * user state is set by the user
+*/
+export interface IOfflineState {
+    readonly isOnline: boolean;
+    readonly rState: eOfflineValue; // runtime state
+    readonly uState: eOfflineValue; // user state
+}
+
+export interface IOfflineListener {
+    isOnline: () => boolean;
+    isListening: () => boolean;
+    unload: () => void;
+    addListener: (callback: OfflineCallback) => IUnloadHook;
+    setOnlineState : (uState: eOfflineValue) => void;
+}
+
+function _disableEvents(target: any, evtNamespace: string | string[]) {
+    eventOff(target, null, null, evtNamespace);
+}
+
+/**
+ * Create a new OfflineListener instance to monitor browser online / offline events
+ * @param parentEvtNamespace - The parent event namespace to append to any specific events for this instance
+ */
+export function createOfflineListener(parentEvtNamespace?: string | string[]): IOfflineListener {
+    let _document = getDocument();
+    var _navigator = getNavigator();        // Gets the window.navigator or workerNavigator depending on the global
+    let _isListening: boolean = false;
+    let listenerList: OfflineCallback[] = [];
+
+    // Set the initial state
+    // rState is changed by the browser, both via events and when we check the navigator.onLine property
+    let rState: eOfflineValue = eOfflineValue.Online;
+    if (_navigator && !isNullOrUndefined(_navigator.onLine) && !_navigator.onLine) { // navigator.onLine is undefined in react-native
+        rState = eOfflineValue.Offline;
+    }
+
+    // ustate is changed by the user calling setOnlineState
+    let uState: eOfflineValue = eOfflineValue.Unknown;
+    // current state would be updated each time rState or uState is changed
+    // it is a resolved value of rState and uState
+    let _currentState: boolean = calCurrentState();
+
+    let _evtNamespace = mergeEvtNamespace(createUniqueNamespace("OfflineListener"), parentEvtNamespace);
+
+    try {
+        if (_enableEvents(getWindow())) {
+            _isListening = true;
+        }
+        
+        if (_document) {
+            // Also attach to the document.body or document
+            let target:any = _document.body || _document;
+
+            if (target.ononline) {
+                if (_enableEvents(target)) {
+                    _isListening = true;
+                }
+            }
+        }
+
+        if (_isListening) {
```

Comment:
I think we can get rid of this now as your setting the `rState` above and we don't want to "reset" the `_currentState` here.